src/libraries/Common/src/SourceGenerators/DiagnosticInfo.cs

```diff
@@ -0,0 +1,44 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Linq;
+using System.Numerics.Hashing;
+using Microsoft.CodeAnalysis;
+
+namespace SourceGenerators;
+
+/// <summary>
+/// Descriptor for diagnostic instances using structural equality comparison.
+/// Provides a work-around for https://github.com/dotnet/roslyn/issues/68291.
+/// </summary>
+internal readonly struct DiagnosticInfo : IEquatable<DiagnosticInfo>
```

Comment:
Likewise, I believe the regex generator defines such a struct as well.

Reply:
At a glance this seems like a type that would be defined as a `record` if arrays were naturally equitable. Why wasn't this just defined as a `record` with `ImmutableEquatableArray<object?>` wrapping `MessageArgs`?

Reply:
It would force an intermediate allocation from `object[]` to `ImmutableEquatableArray` then back to an `object[]` when the `Diagnostic` gets materialized. I think it's small enough to justify a bespoke type. Longer term though we should be looking at addressing https://github.com/dotnet/roslyn/issues/68291 and use `Diagnostic` in the equatable models directly.

Reply:
> Diagnostic will never be ok to use in a model. 

Per https://github.com/dotnet/runtime/pull/89587#discussion_r1334856473 this bit is resolved for this PR.